### DISCUSSION TITLE
Batch with overlap

### DISCUSSION THREAD

**Discussion by matyasbach** [morelinq/MoreLINQ#1005] [Q&A] [answered]
Hi! Is there any simple way to _chunkify_ IEnumerable<T> with some overlap? Or can I propose API like bellow or try to do PR?

```C#
var input = new [] { 1, 2, 3, 4, 5, 6 };
var result = input.Batch(3, overlap: 1); // => [[1, 2, 3], [3, 4, 5], [5, 6]]
```
Basically something between Batch() and Window().

**Comment by viceroypenguin** [comment]
@matyasbach This is available already in [SuperLinq](https://github.com/viceroypenguin/SuperLinq) under the [`Buffer`](https://viceroypenguin.github.io/SuperLinq/api/SuperLinq.SuperEnumerable.Buffer.html#SuperLinq_SuperEnumerable_Buffer__1_System_Collections_Generic_IEnumerable___0__System_Int32_System_Int32_) operator. In this case, you would use `input.Buffer(count: 3, skip: 2);`.

**Comment by atifaziz** [comment, marked as answer]
You can use [`WindowLeft`](https://morelinq.github.io/3.4/ref/api/html/M_MoreLinq_MoreEnumerable_WindowLeft__1.htm) with [`TakeEvery`](https://morelinq.github.io/3.4/ref/api/html/M_MoreLinq_MoreEnumerable_TakeEvery__1.htm) like so:

```c#
var input = new [] { 1, 2, 3, 4, 5, 6 };
var result = input.WindowLeft(3).TakeEvery(2);

Console.WriteLine($"[{result.Select(w => $"[{w.ToDelimitedString(", ")}]")
                            .ToDelimitedString(", ")}]");

// => [[1, 2, 3], [3, 4, 5], [5, 6]]
```

However, creating windows only to skip some may be a bit wasteful so you may want to use [`Buffer`](https://www.fuget.org/packages/System.Interactive/6.0.1/lib/netstandard2.0/System.Interactive.dll/System.Linq/EnumerableEx?code=true#M%3ASystem.Linq.EnumerableEx.Buffer%60%601%28System.Collections.Generic.IEnumerable%7B%60%600%7D%2CSystem.Int32%2CSystem.Int32%29) from [**System.Interactive**](https://www.nuget.org/packages/System.Interactive/) instead.

**Reply by matyasbach** [reply]
Thank you for the answer (and this well-crafted library).

In my use-case I need chunks of 500 items, so taking every 499 window would indeed be very wasteful. Thanks for the hint for System.Interactive, I didn't know this library.

What about the proposal for Batch overload?

**Reply by atifaziz** [reply]
> What about the proposal for Batch overload?

If **System.Interactive** already offers `Buffer`, what would be the point of duplicating the same functionality and maintenance effort in MoreLINQ?

**Reply by matyasbach** [reply]
The point would be having discoverable and intuitive method for this use-case. For example I have not found `Buffer` method and it's _skip_ parameter seems weird to me (what is skipped? Doc must be read to understand it, so it makes code harder to read IMO or custom extension method which would wrap it must be written).

Another point would be not having to depend on 2 libraries (MoreLinq for most of advanced stuff, System.Interactive for this method only) as it's some maintenance effort too.

But I can understand your point of view.
